### PR TITLE
new lints len_zero and len_without_is_empty

### LINKED ISSUE #60: trait with a len() method but without is_empty() method
Regarding my failed attempt to fully solve #32 (which would require more typeck than external lints can currently acquire), I thought about the *social* contract around `.len()` and `.is_empty()`. While we don't have a `Bounded` trait that defines `.len()` and implements `.is_empty()`, we can still ask implementers to provide the latter.

There is some precedence to the use of is_empty: E.g. in Java, the `Collection` interface defines `boolean isEmpty()` with the usual meaning. C++'s STL defines `empty()` on all containers, Ocaml at least defines `is_empty` for stacks. Python's lists and Lua's tables are basically arrays, so they have no problem with using `len(_)`.

In any event, we could add a lint that looks for traits which define `.len()` but do not define `.is_empty()` and ask the implementer to consider adding it. This would also remove the need for method lookup on that other lint – if most traits have `.is_empty()`, we won't need to care for the few remaining false positives, right?

A basic implementation would `check_item` for `ItemTrait`s and see if the contained `TraitItem`s fail to adhere to the social contract.

### PR DESCRIPTION
**Pull request by llogiq**
This closes #60 and at least puts #32 on hold until we can have method lookup.

### CODE REVIEW

**Comment by llogiq** [comment]
@Manishearth Did you change travis to stable? The build is failing. Also my last commit can probably be reverted.

**Comment by Manishearth** [comment]
Travis changed itself to default to stable. I did locally fix this for all my crates, but I had flaky internet and the clippy push probably didn't go through.

File: tests/compile-fail/len_zero.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#![feature(plugin)]
+#![plugin(clippy)]
+
+struct One;
+
+#[deny(len_without_is_empty)]
+impl One {
+	fn len(self: &Self) -> isize { //~ERROR
```
Comment (Manishearth) [inline tests/compile-fail/len_zero.rs]: Nit: keep at least the beginning bits of the error message here (and for most of the ones below too)

**Comment by llogiq** [comment]
I'll close this and switch to another branch.

**Pull request closed without merging**